pallets/confidential/src/lib.rs

```diff
@@ -0,0 +1,169 @@
+
+        #[weight = SimpleDispatchInfo::FixedNormal(400_000)]
+        pub fn add_range_proof(origin,
+            target_id: IdentityId,
+            ticker: Ticker,
+            secret_value: u64,
+        ) -> DispatchResult
+        {
+            let prover_acc = ensure_signed(origin)?;
+            let prover = Context::current_identity_or::<Identity<T>>(&prover_acc)?;
+
+            // Create proof
+            let mut rng = rng::Rng::default();
+            let rand_blind = Scalar::random(&mut rng);
+            let in_range_proof = prove_within_range( secret_value, rand_blind, 32, &mut rng)
+                .map_err(|e| {
+                    debug::error!("Confidential error: {:?}", e);
+                    Error::<T>::InvalidRangeProof
+                })?;
+
+            let ticker_range_proof = TickerRangeProof {
+                initial_message: in_range_proof.init.as_bytes().into(),
+                final_response: in_range_proof.response.to_bytes().into(),
+                max_two_exp: 32,
+            };
+            let prover_ticker_key = ProverTickerKey { prover, ticker };
+            <RangeProofs>::insert(&target_id, &prover_ticker_key, ticker_range_proof);
+            Ok(())
+        }
+
+        #[weight = SimpleDispatchInfo::FixedNormal(400_000)]
+        pub fn add_verify_range_proof(origin,
+            target: IdentityId,
+            prover: IdentityId,
+            ticker: Ticker) -> DispatchResult
+        {
+            let verifier = ensure_signed(origin)?;
+            let verifier_id = Context::current_identity_or::<Identity<T>>(&verifier)?;
+
+            Self::verify_range_proof(target, prover, ticker.clone())?;
+
+            <RangeProofVerifications>::insert((target,ticker), verifier_id, true);
+            Ok(())
+        }
+    }
+}
+
+decl_event! {
+    pub enum Event
+    {
+        RangeProofAdded(IdentityId, Ticker, TickerRangeProof),
+        RangeProofVerified(IdentityId, IdentityId, Ticker),
+    }
+}
+
+decl_error! {
+    pub enum Error for Module<T: Trait> {
+        ///
+        MissingRangeProof,
+        ///
+        InvalidRangeProof,
+    }
+}
+
+impl<T: Trait> Module<T> {
+    pub fn verify_range_proof(
+        target: IdentityId,
+        prover: IdentityId,
+        ticker: Ticker,
+    ) -> DispatchResult {
+        let prover_ticker_key = ProverTickerKey { prover, ticker };
+        let mut rng = rng::Rng::default();
+
+        let trp = Self::range_proof(&target, &prover_ticker_key)
+            .ok_or_else(|| Error::<T>::MissingRangeProof)?;
+
+        let init = CompressedRistretto::from_slice(trp.initial_message.as_slice());
+        let response = RangeProof::from_bytes(trp.final_response.as_slice())
+            .map_err(|_| Error::<T>::InvalidRangeProof)?;
+        let proof = InRangeProof {
+            init,
+            response,
+            range: 32,
+        };
+
+        verify_within_range(&proof, &mut rng).map_err(|_| Error::<T>::InvalidRangeProof.into())
```

Comment:
Since the rng can be different on different nodes, this function can produce different results and different results will cause a chain split. Am I missing something?

Reply:
Verification fn is designed to produce the _same result even if we use different `rng_`. The idea behind that is to ensure that no one could create a "cheated" proof using a well-know 'challenges'. As Verifier uses a RNG to verify the proof, the prover cannot know it previously, and it cannot create a false positive proof. 

Anyway, crypto team is **evaluating a change in the interface** of the API. Instead of using `RNG`, they could **use hash of the inputs, and chain several 'challenges'** (i.e. the second challenge depends on the first challenge, and the first challenge depends on the inputs).

Reply:
> The idea behind that is to ensure that no one could create a "cheated" proof using a well-know 'challenges'.

The problem remains the same. Albeit minuscule, there is a chance that someone crafts a proof that verifies for a certain random number but not for others. If this happens, this will be a consensus error. I am not comfortable having situations where there can be a consensus error. This is why I suggested using the offchain approach earlier where the rng is generated outside consensus and the result is "voted" inside consensus. Since the results will be static after the initial generation and operators will not verify each other's results, there's no risk of a consensus failure.

Reply:
The trick is the number of challenges per proof. I think is almost impossible that someone crafts a proof... anyway @ajivanyan can answer better than me this question :wink: .
Thinking twice it sill harder to cheat because the any MERCAT transaction is composed by several Proofs.
I do not disagree with the `offchain` approach but we don't need it in that early step.

Pls keep in mind that this ticket is just to ensure that Core and Crypto can be integrate at compilation level (not functional) and discover any friction/difficulty on that ... like for instance the RNG complexity in wasm.

Reply:
If a proof verifies for any random number then random numbers are not essential for verification because one could pick and use constants. Or I'm probably missing something...

Reply:
@vkomenda A valid proof does. However, someone can reverse engineer a "fake invalid proof" that verifies for a particular number but not for others.

@fmiguelgarcia In that case, can we simply replace the rng with a constant (or the substrate flip random)? Keep the rng stuff as it is for future use but use a constant for now. Using a constant will make the verification process vulnerable but it will eliminate the chances of a consensus failure. As you said, it's just initial integration and we can work on proper rng in a different ticket. I'd like to avoid any consensus failure conditions.

Reply:
Assuming the validator is honest, there is no consensus error risk associated with the usage of rng in the verification. As this rng is used for efficiently batch multiple proofs, which could also be verified independently and without requiring the extra rng. There is a negligible opportunity one can create proofs which pass the batch verification for a certain challenge but do not pass the single-proof verification process.

Reply:
@ajivanyan I know that probability is negligible, but if it comes to that, would you recommend sampling the challenge space more than once? i.e. run verification with more than one challenges.